File: wick_deep_agent/wick_deep_agent/messages.py
"""Typed message classes and Messages chain for wick_deep_agent.

Mirrors the Go agent.Messages API with a LangChain-style interface::

    from wick_deep_agent.messages import SystemMessage, HumanMessage, Messages

    chain = SystemMessage("You are helpful.") + HumanMessage("Hi!")
    chain.validate()
    chain.to_list()  # [{"role": "system", "content": "..."}, ...]
"""

from __future__ import annotations

import abc
from dataclasses import dataclass, field
from typing import Any

__all__ = [
    "AIMessage",
    "BaseMessage",
    "HumanMessage",
    "Messages",
    "SystemMessage",
    "ToolMessage",
]


# ---------------------------------------------------------------------------
# Role constants
# ---------------------------------------------------------------------------

ROLE_SYSTEM = "system"
ROLE_USER = "user"
ROLE_ASSISTANT = "assistant"
ROLE_TOOL = "tool"

_VALID_ROLES = {ROLE_SYSTEM, ROLE_USER, ROLE_ASSISTANT, ROLE_TOOL}
_USER_INPUT_ROLES = {ROLE_SYSTEM, ROLE_USER}

# ---------------------------------------------------------------------------
# Message classes
# ---------------------------------------------------------------------------


@dataclass
class BaseMessage(abc.ABC):
    """Abstract base chat message. Subclasses must set ``role`` via ``__post_init__``."""

    content: str
    role: str = ""

    def __init_subclass__(cls, **kwargs: Any) -> None:
        super().__init_subclass__(**kwargs)

    def __post_init__(self) -> None:
        if type(self) is BaseMessage:
            raise TypeError("BaseMessage cannot be instantiated directly; use a subclass")

    def to_dict(self) -> dict[str, Any]:
        return {"role": self.role, "content": self.content}

    def validate(self) -> None:
        if self.role not in _VALID_ROLES:
            raise ValueError(f"invalid role '{self.role}': must be one of {sorted(_VALID_ROLES)}")
        if self.role in (ROLE_USER, ROLE_SYSTEM) and not self.content:
            raise ValueError(f"{self.role} message must have non-empty content")

    # Chain operators ---------------------------------------------------------

    def __add__(self, other: "BaseMessage | Messages") -> "Messages":
        if isinstance(other, Messages):
            return Messages(self, *other._msgs)
        if isinstance(other, BaseMessage):
            return Messages(self, other)
        return NotImplemented

    def __radd__(self, other: "BaseMessage | Messages") -> "Messages":
        if isinstance(other, Messages):
            return Messages(*other._msgs, self)
        if isinstance(other, BaseMessage):
            return Messages(other, self)
        return NotImplemented

    def __repr__(self) -> str:
        content_preview = self.content[:60] + "..." if len(self.content) > 60 else self.content
        return f"{self.__class__.__name__}({content_preview!r})"


@dataclass
class SystemMessage(BaseMessage):
    """System-role message."""

    content: str = ""

    def __post_init__(self) -> None:
        self.role = ROLE_SYSTEM


@dataclass
class HumanMessage(BaseMessage):
    """User-role message."""

    content: str = ""

    def __post_init__(self) -> None:
        self.role = ROLE_USER


@dataclass
class AIMessage(BaseMessage):
    """Assistant-role message, optionally carrying tool calls."""

    content: str = ""
    tool_calls: list[dict[str, Any]] = field(default_factory=list)

    def __post_init__(self) -> None:
        self.role = ROLE_ASSISTANT

    def to_dict(self) -> dict[str, Any]:
        d: dict[str, Any] = {"role": self.role, "content": self.content}
        if self.tool_calls:
            d["tool_calls"] = self.tool_calls
        return d

    def validate(self) -> None:
        if self.role not in _VALID_ROLES:
            raise ValueError(f"invalid role '{self.role}'")
        if not self.content and not self.tool_calls:
            raise ValueError("assistant message must have content or tool_calls")


@dataclass
class ToolMessage(BaseMessage):
    """Tool result message."""

    content: str = ""
    tool_call_id: str = ""
    name: str = ""

    def __post_init__(self) -> None:
        self.role = ROLE_TOOL

    def to_dict(self) -> dict[str, Any]:
        return {
            "role": self.role,
            "content": self.content,
            "tool_call_id": self.tool_call_id,
            "name": self.name,
        }

    def validate(self) -> None:
        if not self.tool_call_id:
            raise ValueError("tool message must have tool_call_id")
        if not self.name:
            raise ValueError("tool message must have name")


# ---------------------------------------------------------------------------
# Messages chain
# ---------------------------------------------------------------------------

_ROLE_LABEL = {
    ROLE_SYSTEM: "System",
    ROLE_USER: "Human",
    ROLE_ASSISTANT: "AI",
    ROLE_TOOL: "Tool",
}


class Messages:
    """Ordered chain of chat messages with builder, filter, and validation API.

    Build via positional args::

        chain = Messages(SystemMessage("..."), HumanMessage("..."))

    Or via fluent builder::

        chain = Messages().system("...").human("...").ai("...").tool(...)

    Or via ``+`` operator::

        chain = SystemMessage("...") + HumanMessage("...")
    """

    __slots__ = ("_msgs",)

    def __init__(self, *msgs: BaseMessage) -> None:
        self._msgs: list[BaseMessage] = list(msgs)

    # -- Fluent builder ------------------------------------------------------

    def system(self, content: str) -> Messages:
        self._msgs.append(SystemMessage(content))
        return self

    def human(self, content: str) -> Messages:
        self._msgs.append(HumanMessage(content))
        return self

    def ai(self, content: str, tool_calls: list[dict[str, Any]] | None = None) -> Messages:
        self._msgs.append(AIMessage(content, tool_calls=tool_calls or []))
        return self

    def tool(self, tool_call_id: str, name: str, content: str) -> Messages:
        self._msgs.append(ToolMessage(content=content, tool_call_id=tool_call_id, name=name))
        return self

    def add(self, *msgs: BaseMessage) -> Messages:
        self._msgs.extend(msgs)
        return self

    def concat(self, other: Messages) -> Messages:
        """Return a *new* chain merging self and other."""
        return Messages(*self._msgs, *other._msgs)

    # -- Operators -----------------------------------------------------------

    def __add__(self, other: "BaseMessage | Messages") -> Messages:
        if isinstance(other, Messages):
            return Messages(*self._msgs, *other._msgs)
        if isinstance(other, BaseMessage):
            return Messages(*self._msgs, other)
        return NotImplemented

    def __radd__(self, other: "BaseMessage | Messages") -> Messages:
        if isinstance(other, BaseMessage):
            return Messages(other, *self._msgs)
        return NotImplemented

    # -- Accessors -----------------------------------------------------------

    def last(self) -> BaseMessage | None:
        return self._msgs[-1] if self._msgs else None

    def last_content(self) -> str:
        return self._msgs[-1].content if self._msgs else ""

    def roles(self) -> list[str]:
        return [m.role for m in self._msgs]

    def __len__(self) -> int:
        return len(self._msgs)

    def __iter__(self):
        return iter(self._msgs)

    def __getitem__(self, idx):
        return self._msgs[idx]

    # -- Conversion ----------------------------------------------------------

    def to_list(self) -> list[dict[str, Any]]:
        """Convert to list-of-dicts for API submission."""
        return [m.to_dict() for m in self._msgs]

    # -- Filtering -----------------------------------------------------------

    def _by_role(self, role: str) -> Messages:
        return Messages(*[m for m in self._msgs if m.role == role])

    def system_messages(self) -> Messages:
        return self._by_role(ROLE_SYSTEM)

    def user_messages(self) -> Messages:
        return self._by_role(ROLE_USER)

    def ai_messages(self) -> Messages:
        return self._by_role(ROLE_ASSISTANT)

    def tool_messages(self) -> Messages:
        return self._by_role(ROLE_TOOL)

    # -- Validation ----------------------------------------------------------

    def validate(self) -> None:
        """Validate the full chain (all roles allowed)."""
        if not self._msgs:
            raise ValueError("messages chain is empty")
        for m in self._msgs:
            m.validate()

    def validate_input(self) -> None:
        """Validate for API submission: only user/system roles, non-empty."""
        if not self._msgs:
            raise ValueError("messages list is empty")
        for m in self._msgs:
            if m.role not in _USER_INPUT_ROLES:
                raise ValueError(
                    f"role '{m.role}' not allowed in user input; "
                    f"only {sorted(_USER_INPUT_ROLES)} are accepted"
                )
            if not m.content:
                raise ValueError(f"{m.role} message must have non-empty content")

    # -- Display -------------------------------------------------------------

    def pretty_print(self) -> str:
        lines: list[str] = []
        for m in self._msgs:
            label = _ROLE_LABEL.get(m.role, m.role)
            if m.role == ROLE_TOOL and isinstance(m, ToolMessage):
                label = f"Tool:{m.name}"
            header = f"[{label}]"

            lines.append(header)
            lines.append(m.content)

            if isinstance(m, AIMessage) and m.tool_calls:
                for tc in m.tool_calls:
                    lines.append(f"  -> tool_call: {tc.get('name', '?')}({tc.get('id', '')})")

            lines.append("")
        return "\n".join(lines)

    def __str__(self) -> str:
        return self.pretty_print()

    def __repr__(self) -> str:
        return f"Messages(len={len(self._msgs)}, roles={self.roles()})"

    # -- Token estimation ----------------------------------------------------

    def estimate_tokens(self) -> int:
        """Rough token estimate using len/4 heuristic (matches Go implementation)."""
        total = 0
        for m in self._msgs:
            total += len(m.content) // 4
            if isinstance(m, AIMessage):
                for tc in m.tool_calls:
                    args = tc.get("args", {})
                    if isinstance(args, dict):
                        for v in args.values():
                            total += len(str(v)) // 4
        return max(total, 1) if self._msgs else 0

File: wick_deep_agent/server/agent/messages.go
package agent

import (
	"fmt"
	"strings"
)

// --- Core message types ---

// Message represents a chat message in the conversation.
type Message struct {
	Role       string     `json:"role"`                  // "system", "user", "assistant", "tool"
	Content    string     `json:"content"`
	ToolCalls  []ToolCall `json:"tool_calls,omitempty"`
	ToolCallID string     `json:"tool_call_id,omitempty"` // set when Role == "tool"
	Name       string     `json:"name,omitempty"`         // tool name when Role == "tool"
}

// ToolCall represents an LLM's request to invoke a tool.
type ToolCall struct {
	ID       string         `json:"id"`
	Name     string         `json:"name"`
	Args     map[string]any `json:"args"`
	RawArgs  string         `json:"-"` // raw JSON string from LLM
}

// ToolResult holds the output of a tool execution.
type ToolResult struct {
	ToolCallID string `json:"tool_call_id"`
	Name       string `json:"name"`
	Output     string `json:"output"`
	Error      string `json:"error,omitempty"`
}

// --- Role constants ---

const (
	RoleSystem    = "system"
	RoleUser      = "user"
	RoleAssistant = "assistant"
	RoleTool      = "tool"
)

// ValidRole returns true if r is a known message role.
func ValidRole(r string) bool {
	switch r {
	case RoleSystem, RoleUser, RoleAssistant, RoleTool:
		return true
	}
	return false
}

// UserInputRole returns true if the role is allowed in user-submitted messages.
func UserInputRole(r string) bool {
	return r == RoleUser || r == RoleSystem
}

// --- Constructors (LangChain-style typed message creation) ---

// Human creates a user message.
//
//	Human("hello") → Message{Role:"user", Content:"hello"}
func Human(content string) Message {
	return Message{Role: RoleUser, Content: content}
}

// System creates a system message.
//
//	System("You are a helpful assistant.") → Message{Role:"system", Content:"..."}
func System(content string) Message {
	return Message{Role: RoleSystem, Content: content}
}

// AI creates an assistant message with optional tool calls.
//
//	AI("Sure, I can help.")                      → plain response
//	AI("", tc1, tc2)                             → tool-calling response
func AI(content string, toolCalls ...ToolCall) Message {
	return Message{Role: RoleAssistant, Content: content, ToolCalls: toolCalls}
}

// ToolMsg creates a tool result message.
//
//	ToolMsg("call_123", "calculate", "42")
func ToolMsg(toolCallID, name, output string) Message {
	return Message{Role: RoleTool, Content: output, ToolCallID: toolCallID, Name: name}
}

// --- Messages chain type ---

// Messages is an ordered list of messages with builder methods for chain construction.
//
// Usage:
//
//	chain := NewMessages().
//	    System("You are helpful.").
//	    Human("What is 2+2?")
//
//	// After LLM response:
//	chain = chain.AI("Let me calculate.", calcToolCall).
//	    Tool("call_1", "calculate", "4").
//	    AI("2+2 = 4")
type Messages []Message

// NewMessages creates an empty message chain.
func NewMessages(msgs ...Message) Messages {
	return Messages(msgs)
}

// System appends a system message and returns the chain.
func (m Messages) System(content string) Messages {
	return append(m, System(content))
}

// Human appends a user message and returns the chain.
func (m Messages) Human(content string) Messages {
	return append(m, Human(content))
}

// AI appends an assistant message and returns the chain.
func (m Messages) AI(content string, toolCalls ...ToolCall) Messages {
	return append(m, AI(content, toolCalls...))
}

// Tool appends a tool result message and returns the chain.
func (m Messages) Tool(toolCallID, name, output string) Messages {
	return append(m, ToolMsg(toolCallID, name, output))
}

// Add appends one or more messages and returns the chain.
func (m Messages) Add(msgs ...Message) Messages {
	return append(m, msgs...)
}

// Concat merges another chain onto this one.
func (m Messages) Concat(other Messages) Messages {
	return append(m, other...)
}

// Last returns the last message, or a zero Message if empty.
func (m Messages) Last() Message {
	if len(m) == 0 {
		return Message{}
	}
	return m[len(m)-1]
}

// LastContent returns the content of the last message.
func (m Messages) LastContent() string {
	return m.Last().Content
}

// Len returns the number of messages.
func (m Messages) Len() int {
	return len(m)
}

// Slice returns the underlying []Message.
func (m Messages) Slice() []Message {
	return []Message(m)
}

// --- Filtering ---

// ByRole returns messages with the given role.
func (m Messages) ByRole(role string) Messages {
	var out Messages
	for _, msg := range m {
		if msg.Role == role {
			out = append(out, msg)
		}
	}
	return out
}

// UserMessages returns only user messages.
func (m Messages) UserMessages() Messages { return m.ByRole(RoleUser) }

// AssistantMessages returns only assistant messages.
func (m Messages) AssistantMessages() Messages { return m.ByRole(RoleAssistant) }

// ToolMessages returns only tool result messages.
func (m Messages) ToolMessages() Messages { return m.ByRole(RoleTool) }

// SystemMessages returns only system messages.
func (m Messages) SystemMessages() Messages { return m.ByRole(RoleSystem) }

// --- Validation ---

// Validate checks that the message chain is well-formed:
//   - All roles are valid
//   - Tool messages have ToolCallID and Name set
//   - Assistant messages with ToolCalls have non-empty call IDs
//   - No empty content (except assistant messages with tool calls)
func (m Messages) Validate() error {
	for i, msg := range m {
		if !ValidRole(msg.Role) {
			return fmt.Errorf("message[%d]: unknown role %q", i, msg.Role)
		}

		switch msg.Role {
		case RoleTool:
			if msg.ToolCallID == "" {
				return fmt.Errorf("message[%d]: tool message missing tool_call_id", i)
			}
			if msg.Name == "" {
				return fmt.Errorf("message[%d]: tool message missing name", i)
			}

		case RoleAssistant:
			// Assistant messages can have empty content if they have tool calls
			if msg.Content == "" && len(msg.ToolCalls) == 0 {
				return fmt.Errorf("message[%d]: assistant message has no content and no tool calls", i)
			}
			for j, tc := range msg.ToolCalls {
				if tc.ID == "" {
					return fmt.Errorf("message[%d].tool_calls[%d]: missing ID", i, j)
				}
				if tc.Name == "" {
					return fmt.Errorf("message[%d].tool_calls[%d]: missing name", i, j)
				}
			}

		case RoleUser, RoleSystem:
			if msg.Content == "" {
				return fmt.Errorf("message[%d]: %s message has empty content", i, msg.Role)
			}
		}
	}
	return nil
}

// ValidateUserInput checks that messages are valid for user submission
// (only user and system roles allowed).
func (m Messages) ValidateUserInput() error {
	if len(m) == 0 {
		return fmt.Errorf("messages must not be empty")
	}
	for i, msg := range m {
		if !UserInputRole(msg.Role) {
			return fmt.Errorf("message[%d]: role %q not allowed (must be \"user\" or \"system\")", i, msg.Role)
		}
		if msg.Content == "" {
			return fmt.Errorf("message[%d]: content must not be empty", i)
		}
	}
	return nil
}

// --- Display ---

// PrettyPrint returns a human-readable representation of the message chain.
func (m Messages) PrettyPrint() string {
	var sb strings.Builder
	for _, msg := range m {
		sb.WriteString(prettyMessage(msg))
		sb.WriteString("\n")
	}
	return sb.String()
}

// String implements fmt.Stringer.
func (m Messages) String() string {
	return m.PrettyPrint()
}

func prettyMessage(msg Message) string {
	var sb strings.Builder
	label := roleLabel(msg.Role)

	if msg.Role == RoleTool {
		sb.WriteString(fmt.Sprintf("[%s: %s (call_id=%s)]\n", label, msg.Name, msg.ToolCallID))
	} else {
		sb.WriteString(fmt.Sprintf("[%s]\n", label))
	}

	if msg.Content != "" {
		sb.WriteString(msg.Content)
		sb.WriteString("\n")
	}

	for _, tc := range msg.ToolCalls {
		sb.WriteString(fmt.Sprintf("  → tool_call: %s(id=%s, args=%v)\n", tc.Name, tc.ID, tc.Args))
	}

	return sb.String()
}

func roleLabel(role string) string {
	switch role {
	case RoleSystem:
		return "System"
	case RoleUser:
		return "Human"
	case RoleAssistant:
		return "AI"
	case RoleTool:
		return "Tool"
	default:
		return role
	}
}

// --- Token estimation ---

// EstimateTokens returns a rough token count (len/4 heuristic).
func (m Messages) EstimateTokens() int {
	total := 0
	for _, msg := range m {
		total += len(msg.Content) / 4
		for _, tc := range msg.ToolCalls {
			total += len(tc.RawArgs) / 4
		}
	}
	return total
}
